### PR TITLE
[Substrait to Velox] Add conversion for hash join

### PR DESCRIPTION
**Pull request by marin-ma**
Add substrait-to-velox conversion for JoinRel. Since substrait doesn't distinguish between different kinds of physical join types, we'll convert JoinRel to hash join by default.

This PR depends on https://github.com/facebookincubator/velox/pull/1554. Will rebase after https://github.com/facebookincubator/velox/pull/1554 gets merged.

### CODE REVIEW

File: velox/substrait/SubstraitToVeloxPlan.h [hunk 1]
```diff
@@ -40,7 +40,11 @@ class SubstraitVeloxPlanConverter {
       const ::substrait::FilterRel& filterRel,
       memory::MemoryPool* pool);
 
-  /// Convert Substrait ReadRel into Velox PlanNode.
+  /// Used to convert Substrait JoinRel into Velox PlanNode.
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxPlan.h:43]: nit: "Convert JoinRel into Velox HashJoinNode."

File: velox/substrait/SubstraitToVeloxPlan.cpp [hunk 2]
```diff
@@ -228,6 +228,95 @@ SubstraitVeloxPlanConverter::toVeloxAggWithRowConstruct(
...
+    VELOX_FAIL("Right Rel is expected in JoinRel.");
+  }
+
+  auto leftNode = toVeloxPlan(sJoin.left(), pool);
+  auto rightNode = toVeloxPlan(sJoin.right(), pool);
+
+  auto outputSize =
+      leftNode->outputType()->size() + rightNode->outputType()->size();
+  std::vector<std::string> outputNames;
+  std::vector<std::shared_ptr<const Type>> outputTypes;
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxPlan.cpp:247]: use TypePtr for readability

File: velox/substrait/SubstraitToVeloxPlan.cpp [hunk 3]
```diff
@@ -228,6 +228,95 @@ SubstraitVeloxPlanConverter::toVeloxAggWithRowConstruct(
...
+  for (const auto& node : {leftNode, rightNode}) {
+    const auto& names = node->outputType()->names();
+    outputNames.insert(outputNames.end(), names.begin(), names.end());
+    const auto& types = node->outputType()->children();
+    outputTypes.insert(outputTypes.end(), types.begin(), types.end());
+  }
+  auto outputRowType = std::make_shared<const RowType>(
+      std::move(outputNames), std::move(outputTypes));
+
+  // extract join keys from join expression
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxPlan.cpp:259]: Comments need to be full sentences, e.g. start with a capital letter and end with a period.

File: velox/substrait/SubstraitToVeloxPlan.cpp [hunk 4]
```diff
@@ -228,6 +228,95 @@ SubstraitVeloxPlanConverter::toVeloxAggWithRowConstruct(
...
+  std::vector<const ::substrait::Expression::FieldReference*> leftExprs,
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxPlan.cpp:260]: Declare each variable separately.

File: velox/substrait/SubstraitToVeloxPlan.cpp [hunk 5]
```diff
@@ -228,6 +228,95 @@ SubstraitVeloxPlanConverter::toVeloxAggWithRowConstruct(
...
+      rightExprs;
+  extractJoinKeys(sJoin.expression(), leftExprs, rightExprs);
+  VELOX_CHECK_EQ(leftExprs.size(), rightExprs.size());
+  size_t numKeys = leftExprs.size();
+
+  std::vector<std::shared_ptr<const core::FieldAccessTypedExpr>> leftKeys,
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxPlan.cpp:266]: same

File: velox/substrait/SubstraitToVeloxPlan.cpp [hunk 6]
```diff
@@ -228,6 +228,95 @@ SubstraitVeloxPlanConverter::toVeloxAggWithRowConstruct(
...
+  leftKeys.reserve(numKeys);
+  rightKeys.reserve(numKeys);
+  for (size_t i = 0; i < numKeys; ++i) {
+    leftKeys.emplace_back(
+        exprConverter_->toVeloxExpr(*leftExprs[i], outputRowType));
+    rightKeys.emplace_back(
+        exprConverter_->toVeloxExpr(*rightExprs[i], outputRowType));
+  }
+
+  std::shared_ptr<const core::ITypedExpr> filter;
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxPlan.cpp:277]: use TypedExprPtr for readability

File: velox/substrait/SubstraitToVeloxPlan.cpp [hunk 7]
```diff
@@ -228,6 +228,95 @@ SubstraitVeloxPlanConverter::toVeloxAggWithRowConstruct(
...
+  if (sJoin.has_post_join_filter()) {
+    filter =
+        exprConverter_->toVeloxExpr(sJoin.post_join_filter(), outputRowType);
+  }
+
+  // Map join type
+  core::JoinType joinType;
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxPlan.cpp:284]: nit: consider extracting a helper method to keep the length of this method within reason.

File: velox/substrait/SubstraitToVeloxPlan.cpp [hunk 8]
```diff
@@ -812,4 +904,38 @@ int32_t SubstraitVeloxPlanConverter::streamIsInput(
...
+  while (!expressions.empty()) {
+    auto visited = expressions.back();
+    expressions.pop_back();
+    if (visited->rex_type_case() ==
+        ::substrait::Expression::RexTypeCase::kScalarFunction) {
+      auto functionSpec = substraitParser_->findFunctionSpec(
+          functionMap_, visited->scalar_function().function_reference());
+      auto functionName = substraitParser_->getFunctionName(functionSpec);
+      const auto& args = visited->scalar_function().args();
+      if (functionName == "and") {
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxPlan.cpp:922]: nit: use constants

```
static const std::string kAnd = "and";
static const std::string kEqual = "equal";
```

**Review by mbasmanova — COMMENTED** [review]
@marin-ma The second commit looks, but it needs a test. I suggest to add conversion from Velox to substrait as well and then extend the existing round-trip test.

**Comment by marin-ma** [comment]
> @marin-ma The second commit looks, but it needs a test. I suggest to add conversion from Velox to substrait as well and then extend the existing round-trip test.

@ZJie1 is working on conversion from Velox to substrait. She will raise a PR for this part.

**Comment by mbasmanova** [comment]
> substrait doesn't distinguish between different kinds of physical join types

Do you know if there are plan to support different join types? Should we open an issue with Substrait? CC: @jacques-n

**Comment by marin-ma** [comment]
> > substrait doesn't distinguish between different kinds of physical join types
> 
> Do you know if there are plan to support different join types? Should we open an issue with Substrait? CC: @jacques-n

I've asked in this issue before https://github.com/substrait-io/substrait/issues/189
It's in their plan, but we don't have the people to do it right now.

**Comment by mbasmanova** [comment]
@marin-ma Rong, thank you for clarifying. It seems to be that this PR is blocked on #1554 and on additional work to translate join node from Velox to Substrait + test. Let's work on #1554 then.

**Comment by jacques-n** [comment]
For reference, the related Substrait issue should be relatively simple rote work (since it is simply writing a proto for the already defined specification).

**Comment by rui-mo** [comment]
Looks like this PR is still needed. @zhejiangxiaomai

**Pull request closed without merging**